File: typescript/grains/grains.ts
type SquareNumber = number;

function isSquareNumber(n: SquareNumber): n is SquareNumber {
  return n > 0 && n <= 64;
}

class Grains {
  static total(): number {
    return Array.from(Array(65).keys()).reduce(
      (sum, squareNumber) => sum + this.square(squareNumber)
    );
  }
  static square(squareNumber: number): any {
    if (isSquareNumber(squareNumber)) {
      return Math.pow(2, squareNumber - 1);
    } else {
      throw new Error("Input has wrong format");
    }
  }
}

export default Grains;

File: elixir/grains/lib/grains.ex
defmodule Grains do
  @doc """
  Calculate two to the power of the input minus one.
  """
  @spec square(pos_integer()) :: {:ok, pos_integer()} | {:error, String.t()}
  def square(square) when square > 64 or square < 1, do: {:error, "The requested square must be between 1 and 64 (inclusive)"}
  def square(number), do: {:ok, grains_for(number)}

  defp grains_for(square) when square in [1, 2], do: square
  defp grains_for(square), do: power_of(2, square - 1)

  defp power_of(number, 1), do: number
  defp power_of(number, exp), do: number * power_of(number, exp - 1)

  @doc """
  Adds square of each number from 1 to 64.
  """
  @spec total :: {:ok, pos_integer()}
  def total do
    {:ok, Enum.reduce(1..64, fn square, total -> total + grains_for(square) end)}
  end
end
